composer/trainer/dist_strategy.py

```diff
@@ -223,6 +223,16 @@ def sync_hook(*args):
     state_dict_type = fsdp_config.get('state_dict_type', 'full')
     activation_checkpointing_reentrant = fsdp_config.get('activation_checkpointing_reentrant', True)
 
+    kwargs = {}
+    if is_torch_2_0:
+        # Support of new parameter `use_orig_params` in PyTorch 2.0 or higher.
+        # Setting this to `True` has FSDP use `module`'s original parameters via method
+        # `nn.Module.named_parameters` instead of FSDP's internal class `FlatParameter`. However,
+        # setting it to `False` exposes FSDP's internal class `FlatParameter` via method
+        # `nn.Module.named_parameters`.
+        # Setting it to `True` is mandatory when using `torch.compile()`.
+        kwargs['use_orig_params'] = fsdp_config.get('use_orig_params', True)
```

Comment:
what error does the user face if using FSDP + torch2 and setting `use_orig_params` to `False`? Should we error here in that case with an informative error?

Reply:
- `FSDP` + `torch 2.0` + `use_orig_params=False` = no error
- `FSDP` + `torch 2.0` + `use_orig_params=True` = no error
- `FSDP` + `torch 2.0` + `use_orig_params=True` + without `torch.compile` = no error
- `FSDP` + `torch 2.0` + `use_orig_params=False` + with `torch.compile` = `AssertionError: Dynamo only supports FSDP with use_orig_params=True`

There is no harm in keeping `use_orig_params` to `True` for `torch 2.0` irrespective of torch.compile or not. If we don't do that then it would require a little complex logic check here where we have to check if the provided model by the user is compiled or not or if composer compile the model then we have to pass some sort of variable in `dist_strategy.py` to let `prepare_fsdp_module()` know about this. Hence, we enables it by default for torch 2.0